_posts/2020-04-03-what-does-appsync-do.md

```diff
@@ -0,0 +1,200 @@
+---
+layout: post
+title: What does AWS AppSync do ?
+date: 2020-04-03
+tags: AWS AppSync Cloudformation
+author: Shermayne Lee
+---
+
+# Overview
+
+AWS AppSync and GraphQL have gained popularity in recent years. However it can be confusing and difficult to start for developers new to graphQL concept.
+
+AWS AppSync is a managed GraphQL platform by AWS for serverless implementation. It can translate graphQL requests from client, get the data from AWS Resources and pass the data to the client.
+
+# Introduction to AppSync
+
+AWS Appsync is the service that helps your application development process simple by providing the ability to create flexible API. It can be the central hub that connects all your data sources securely.
+
+AWS AppSync is designed mainly for the use of GraphQL API and one of the main benefits of AppSync is letting your application get the exact data that you need. Additionally for mobile and web applications, it provides the ability to access data locally when devices go offline and synchronize application data in real-time when devices are back online.
+
+# Quick Start
+
+The nice thing about using AWS services is we can create resources with CloudFormation template. It allows you to create and provision AWS resources predictably and repeatedly.
+
+### Step 1: Create AWS AppSync API
+
+We will build an API for a coffee ordering application that store all the order in the database.
+
+The following example is written in YAML format.
+
+```yml
+Resources:
+  CoffeeShopAppSyncApi:
+    Type: AWS::AppSync::GraphQLApi
+    Properties:
+      Name: CoffeeShopAppSyncApi
+      AuthenticationType: API_Key
+
+  CoffeeShopAppSyncApiKey:
+    Type: AWS::AppSync::ApiKey
+    Properties:
+      ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+      Description: API key for graphQL Api
+```
+
+There are two resources in the template.
+
+Firstly, it is creating the resource for AppSyncAPI. There are few methods available for authentication type such as
+
+- API_KEY
+- AWS_IAM
+- AMAZON_COGNITO_USER_POOLS
+- OPENID_CONNECT.
+
+In this case, we are using an API key to protect the API, so we will need to create an API key.
+
+### Step 2: Create the Schema
+
+Now we have the API ready, we will need to create AppSync schema for the API
+
+```yml
+CoffeeShopAppSyncSchema:
+  Type: AWS::AppSync::GraphQLSchema
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    Definition: |
+      input orderInput {
+        id: String
+        coffee: String
+        size: String
+      }
+
+      type Order {
+        id: String
+        coffee: String
+        size: String
+      }
+
+      type Mutation {
+      createOrder(input: orderInput): Order
+      }
+
+      type Query{
+      getOrder(id: String): Order
+      }
+
+      schema {
+      query: Query
+      mutation: Mutation
+      }
+```
+
+The schema that we created serves the purpose of ensuring the data from the client matches the schema before mapping to the resolver.
```

Comment:
"The schema that we created serves the purpose of ensuring the data from the client matches the schema before mapping to the resolver." -> "The schema we have created forms a contract between our clients and the AWS data sources ensuring the data sent back and forth adheres to the specified types."